[Issue 4175] [pulsar-function-go] Add Go Function heartbeat (and gRPC service) for production usage

devinbost opened the pull request:
Partial fix required for #4175. 
Addresses issue documented here: https://github.com/grpc/grpc-go/issues/3310
Part of ongoing work to add the gRPC service for heartbeat functionality and related support for running Go functions. 

### Motivation
Progress was blocked by the issue documented above. 

### Modifications

1. Updated ./generate.sh script in pulsar-function-go module to use the grpc plugin. 
2. Rebuilt Go gRPC files using the grpc plugin. This updated the proto version and was mostly an additive change that included adding the new registration methods required for creating the Go gRPC service. 

### Verifying this change

- Unable to test changes because Go tests were already failing. Currently investigating. 
- More testing will be done in subsequent commits because the Go gRPC functionality is not yet complete. (Should be low impact.)

devinbost commented:
Let's not merge this until I can get the Go tests to pass. I just wanted to get this upstream to share the fix with anyone else who might be working on this feature.

devinbost commented:
FYI, I ran the Go tests from master (without my changes), and they failed with the same results as the tests with the changes from this PR.

wolfstudy commented:
> FYI, I ran the Go tests from master (without my changes), and they failed with the same results as the tests with the changes from this PR.

Thanks @devinbost Can you  provide the error information? Or how can I reproduce this problem

devinbost commented:
@wolfstudy Thanks for taking a look. 
If you clone the latest from my go-features branch on my devinbost/pulsar fork (to `/Users/[yourUserName]/github.com/apache/pulsar/`), then:
`cd /Users/[yourUserName]/github.com/apache/pulsar/pulsar-function-go/pf`
`go test`

You should get this result:

`OCPC-LM31977:pf dbost$ go test`
> # github.com/apache/pulsar/pulsar-function-go/pf [github.com/apache/pulsar/pulsar-function-go/pf.test]
> ./function.go:167:44: undefined: port
> FAIL    github.com/apache/pulsar/pulsar-function-go/pf [build failed]
> OCPC-LM31977:pf dbost$ go test
> --- FAIL: TestContext (0.00s)
>     context_test.go:38: 
>                 Error Trace:    context_test.go:38
>                 Error:          Not equal: 
>                                 expected: []string{"topic-1", "topic-2"}
>                                 actual  : []string(nil)
>                                 
>                                 Diff:
>                                 --- Expected
>                                 +++ Actual
>                                 @@ -1,5 +1,2 @@
>                                 -([]string) (len=2) {
>                                 - (string) (len=7) "topic-1",
>                                 - (string) (len=7) "topic-2"
>                                 -}
>                                 +([]string) <nil>
>                                  
>                 Test:           TestContext
>     context_test.go:42: 
>                 Error Trace:    context_test.go:42
>                 Error:          Not equal: 
>                                 expected: "topic-3"
>                                 actual  : "persistent://public/default/topic-02"
>                                 
>                                 Diff:
>                                 --- Expected
>                                 +++ Actual
>                                 @@ -1 +1 @@
>                                 -topic-3
>                                 +persistent://public/default/topic-02
>                 Test:           TestContext
> 2020/01/11 11:16:50.741 log.go:76: [error] process function error:[function is nil]
> 
> 2020/01/11 11:16:50.741 log.go:76: [error] process function error:[function kind struct is not func]
> 
> 2020/01/11 11:16:50.741 log.go:76: [error] process function error:[functions may not take more than two arguments, but function takes 3]
> 
> 2020/01/11 11:16:50.741 log.go:76: [error] process function error:[function takes two arguments, but the first is not Context. got string]
> 
> 2020/01/11 11:16:50.741 log.go:76: [error] process function error:[function may not return more than two values]
> 
> 2020/01/11 11:16:50.741 log.go:76: [error] process function error:[function returns two values, but the second does not implement error]
> 
> 2020/01/11 11:16:50.742 log.go:76: [error] process function error:[function returns a single value, but it does not implement error]
> 
> 2020/01/11 11:16:50.744 log.go:76: [error] process function error:[bad stuff]
> 
> 2020/01/11 11:16:50.744 log.go:76: [error] process function error:[bad stuff]
> 
> 2020/01/11 11:16:50.744 log.go:76: [error] process function error:[bad stuff]
> 
> 2020/01/11 11:16:50.744 log.go:76: [error] process function error:[bad stuff]
> 
> FAIL
> exit status 1
> FAIL    github.com/apache/pulsar/pulsar-function-go/pf  0.123s
>

sijie commented:
ping @wolfstudy ?

devinbost commented:
I ran each of the Go tests individually. Here are the results:

```
util_test.go passed.
instanceConf_test.go passed.
In function_test.go:
    TestInvalidFunctions passed.
    TestInvokes passed.
In context_test.go:
    TestContext failed. 
```

So, most of them passed. It’s only context_test.go that failed.

Regarding the TestContext failure, here's what I got from my breakpoint:

<img width="999" alt="Screen Shot 2020-01-13 at 8 39 50 AM" src="https://user-images.githubusercontent.com/7418031/72276446-24414500-35ed-11ea-9a27-22e93749e813.png">

devinbost commented:
It turns out that the issue in the `context_test.go` test was in `FunctionContext.GetInputTopics()`. 
The instance method gets its valued populated by `goInstance.setupConsumer()`, which must not have been getting run in the test. I changed the getter on the functionContext to get the inputTopic values directly from `functionContext.instanceConf.funcDetails.GetSource().InputSpecs`. 
We may want to remove dependencies on the functionContext.inputTopics field (and remove the field altogether) to prevent confusion later on. 
Also, the conf.yaml values must have been changed, so I updated the test to match the values that were being provided. 
It's passing locally for me. 

I also added a new test file `instanceControlServicer_test.go` that tests that a gRPC server can be created and successfully connected to. 

We probably still need to test the entire thing end-to-end to ensure that concurrency issues weren't introduced anywhere and that I'm starting the servicer in the correct location.

wolfstudy commented:
> It turns out that the issue in the `context_test.go` test was in `FunctionContext.GetInputTopics()`.
> The instance method gets its valued populated by `goInstance.setupConsumer()`, which must not have been getting run in the test. I changed the getter on the functionContext to get the inputTopic values directly from `functionContext.instanceConf.funcDetails.GetSource().InputSpecs`.
> We may want to remove dependencies on the functionContext.inputTopics field (and remove the field altogether) to prevent confusion later on.
> Also, the conf.yaml values must have been changed, so I updated the test to match the values that were being provided.
> It's passing locally for me.
> 
> I also added a new test file `instanceControlServicer_test.go` that tests that a gRPC server can be created and successfully connected to.
> 
> We probably still need to test the entire thing end-to-end to ensure that concurrency issues weren't introduced anywhere and that I'm starting the servicer in the correct location.

Thanks @devinbost agree with you, maybe we can remove `inputTopics` filed of `FunctionContext` . Can you help to fix the `context_test.go` file?

devinbost commented:
> Thanks @devinbost agree with you, maybe we can remove `inputTopics` filed of `FunctionContext` . Can you help to fix the `context_test.go` file?

Done.

devinbost commented:
Instrumenting Prometheus might be more work than anticipated.

I ran into a blocking issue due to missing features in the Go client of Prometheus and got some pushback from one of their maintainers when I asked about getting those features.
I'm discussing the issue with them, but we will likely need to work around the issue since they don't seem very eager to add the desired functionality. 
For reference, see the discussion here: https://groups.google.com/d/msgid/prometheus-users/3480a316-748c-4351-8a15-1ac88971fd24%40googlegroups.com?utm_medium=email&utm_source=footer

devinbost commented:
So, we probably need to have the Prometheus wiring in a different PR due to its complexity and the larger changes that will be required.

wolfstudy commented on pulsar-function-go/pb/doc.go:
```diff
@@ -17,18 +17,4 @@
-//
-// Files generated by the protoc-gen-go program should not be modified.
-package pb
+package api
```
please don't remove the comments

wolfstudy commented on pulsar-function-go/pf/context.go:
```diff
@@ -43,7 +43,15 @@ func (c *FunctionContext) GetInstanceID() int {
 
 func (c *FunctionContext) GetInputTopics() []string {
-	return c.inputTopics
+	myMap := c.instanceConf.funcDetails.GetSource().InputSpecs
```
```suggestion
	inputMap := c.instanceConf.funcDetails.GetSource().InputSpecs
```

wolfstudy commented on pulsar-function-go/pf/context.go:
```diff
@@ -43,7 +43,15 @@ func (c *FunctionContext) GetInstanceID() int {
+		i++
+	}
+	return inputTopics
+	//return c.inputTopics
```
please remove the comments

wolfstudy commented on pulsar-function-go/pf/instance.go:
```diff
@@ -21,12 +21,13 @@ package pf
 	"github.com/apache/pulsar/pulsar-client-go/pulsar"
 	log "github.com/apache/pulsar/pulsar-function-go/logutil"
-	"github.com/apache/pulsar/pulsar-function-go/pb"
+	pb "github.com/apache/pulsar/pulsar-function-go/pb"
```
```suggestion
	"github.com/apache/pulsar/pulsar-function-go/pb"
```

devinbost replied:
When I make this change, I can't get it to build. I get unresolved variables, even if I remove pb from the references in the code.

wolfstudy commented on pulsar-function-go/pf/instance.go:
```diff
@@ -35,19 +36,57 @@ type goInstance struct {
+	maxIdleTime := gi.context.GetMaxIdleTime()
+	timeSinceLastCheck := now.UnixNano() - gi.lastHealthCheckTs
+	if (timeSinceLastCheck) > (maxIdleTime)  {
+		log.Errorf("Haven't received health check from spawner in a while. Stopping instance...")
```
```suggestion
		log.Error("Haven't received health check from spawner in a while. Stopping instance...")
```

wolfstudy commented on pulsar-function-go/pf/instance.go:
```diff
@@ -35,19 +36,57 @@ type goInstance struct {
+	if (timeSinceLastCheck) > (maxIdleTime)  {
+		log.Errorf("Haven't received health check from spawner in a while. Stopping instance...")
+		gi.close()
+		// os.Exit(1)
```
please remove the commets

wolfstudy commented on pulsar-function-go/pf/instanceConf.go:
```diff
@@ -24,7 +24,7 @@ import (
 
 	"github.com/apache/pulsar/pulsar-function-go/conf"
-	"github.com/apache/pulsar/pulsar-function-go/pb"
+	pb "github.com/apache/pulsar/pulsar-function-go/pb"
```
```suggestion
	"github.com/apache/pulsar/pulsar-function-go/pb"
```

devinbost replied:
Again, when I make this change, I can't get it to build.

wolfstudy commented on pulsar-function-go/pf/instanceControlServicer.go:
```diff
@@ -0,0 +1,71 @@
+	"net"
+	"google.golang.org/grpc"
+	log "github.com/apache/pulsar/pulsar-function-go/logutil"
+	pb "github.com/apache/pulsar/pulsar-function-go/pb"
```
```suggestion
	"github.com/apache/pulsar/pulsar-function-go/pb"
```

devinbost replied:
Again, when I make this change, I can't get it to build.

wolfstudy commented on pulsar-function-go/pf/instance_test.go:
```diff
@@ -0,0 +1,93 @@
+
+
+func testProcessSpawnerHealthCheckTimer(tkr *time.Ticker, lastHealthCheckTs int64, expectedHealthCheckInterval int32, counter *int ){
+	println("Starting processSpawnerHealthCheckTimer")
```
```suggestion
	fmt.Println("Starting processSpawnerHealthCheckTimer")
```

wolfstudy replied:
@devinbost please replace `println()` with `fmt.Println()`

wolfstudy commented on pulsar-function-go/pf/instance_test.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+
+func testProcessSpawnerHealthCheckTimer(tkr *time.Ticker, lastHealthCheckTs int64, expectedHealthCheckInterval int32, counter *int ){
```
```suggestion
func testProcessSpawnerHealthCheckTimer(tkr *time.Ticker, lastHealthCheckTs int64, expectedHealthCheckInterval int32, counter int ){
```

devinbost replied:
This change breaks the test because the counter variable doesn't propagate as needed for the assertion.

wolfstudy commented on pulsar-function-go/pf/instance_test.go:
```diff
@@ -0,0 +1,93 @@
+	}
+}
+
+func testStartScheduler(counter *int){
```
```suggestion
func testStartScheduler(counter int){
```

devinbost replied:
Again, this change breaks the test because the counter variable doesn't propagate as needed for the assertion.

wolfstudy commented:
@devinbost I'm adding Action CI for Go Functions, make ensure that test cases can pass and code style is good

devinbost commented:
I made the changes that I could. Regarding everything that wasn't modified, please see my replies to the comments.

wolfstudy commented:
retest this please

sijie commented:
@devinbost can you rebase this pull request?

devinbost commented:
@sijie It's rebased and ready to go.

wolfstudy commented:
run java8 tests
run integration tests
run cpp tests

wolfstudy commented:
run integration tests
run java8 tests

wolfstudy commented:
run java8 tests

wolfstudy commented:
run integration tests

wolfstudy commented:
run cpp tests

sijie commented:
run java8 tests

devinbost commented:
@sijie Do we need all of the Github Action tests to pass before this can be merged? 
All of the tests pass on my local machine.

devinbost commented:
@merlimat @jerrypeng How do we get this PR merged? We've tried re-running the tests at least 12 times, and the intermittent test failures continue to prevent the tests from all passing.

sijie commented:
@devinbost the actions are marked with required. we can't merge it right now. working on how to get around the github action issues.

devinbost commented:
@sijie Thanks for the help. I really appreciate it.
